docs/content/sips/007-deployment-auth.md

```diff
@@ -0,0 +1,158 @@
+> * Provide a way for the user to say which URL is their default
+> * Require the user to provide a URL
+> * Provide an abstraction (e.g. name, config file path) by which the user can switch between servers
+
+> QUESTION: The last section represents a change in behaviour when no flags are provided. Are we happy with this, or should we direct people to `spin login` and not attempt it automatically?
+
+> QUESTION: Is it correct to reject a username/password if the user is logged in to the chosen server, or should we attempt username/password login in this case?
+
+> QUESTION: What happens if the user is logged in to one server and `$HIPPO_URL` is set to another?  What happens if the user is logged in but the `$HIPPO_USERNAME` is set?  (Thinking here of situations where the user is switching between a username/password server and a device-auth one.)  Should we provide a way to "log into" username/password servers, and use that stored state instead of the environment variables?
+
+## Error flows
+
+### Auth failure on deploy
+
+Deployment would fail if the token was incorrect or expired.
+
+> QUESTION: In this case should we simply fail, or should we attempt login?  _Note:_ if we decide to attempt login we _must_ provide a flag to suppress this (for programmatic consumers who have their own login flows).
+
+It would be useful to have an unambiguous way of reporting this for programmatic consumers (e.g. distinguished error code, JSON output option), so they could display login UI or provide a specific prompt.  That may be a bit much to manage in the first drop though.
+
+## `spin login` detailed considerations
+
+`spin login` is a new command and a relatively complex user experience.  This section describes the behaviour in more detail.
+
+### Browser interaction
+
+As part of device auth flow, the user needs to go to the verification URL in their browser and paste in the code.
+
+> QUESTION: Should we attempt to open the browser automatically?
+> * Pro: it's more convenient
+> * Con: it may be unexpected (we have not asked for consent or warned it will happen)
+> * Con: it may work differently across browsers and operating systems, which may require explanation in the documentation
+
+> QUESTION: Should we copy the user code to the clipboard?
+> * Pro: it's convenient
+> * Con: the user may already have stuff on their clipboard
+> * Con: again, documentation needs to consider differences across OSes
+
+(For both of these questions, is there prior art to which we can refer?)
+
+### Verification polling
+
+The device auth flow involves the client polling the server using the device code as the auth token. Each poll can have three outcomes:
+
+* 200 and the response contains a token: user has authorised this device
+* 200 and the response doesn't contain a token: user has not yet authorised this device
+* Server error
+* Network error
+
+We propose to poll for two minutes. If the user has not entered the code by then, `spin login` fails and exits.
```

Comment:
A device code is provided with an expiration date. The expiration date determines how long the user can take to enter the device code until the code expires. The server is the one that controls how long the device code is valid.

We could check the expiration date to determine how long we can poll for. I believe it's currently available for 900 seconds (15 minutes).

https://github.com/fermyon/cloud-openapi/blob/896a21187ff72a1e779f13be47e8a1685833df85/clients/rust/src/models/device_code_item.rs#L23